### PR TITLE
Hs/updateuser.improvements

### LINKED ISSUE #56: Nicks aren't carried over until a nick-name change
Faux matrix user's display names don't reflect their nicks until they're changed. Should grab their nick initially rather than their discord usernames.

### PR DESCRIPTION
**Pull request by Half-Shot**
This change removes the UpdateUser function and instead spreads out that functionality to a UserSync class where we make the algorithm more verbose and easier to test. 

What does this do:

* Users are now joined and left from the guilds appropriately
* Profile updates are made, and guild nicks overwrite them automatically.
* Pulled out more code from ``bot.ts``

Fixes #56, #101 and #102

### CODE REVIEW

File: test/test_usersyncroniser.ts [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+               expect(LINK_RM_USER).is.not.null;
+               expect(AVATAR_SET).is.null;
+               expect(UTIL_UPLOADED_AVATAR).to.be.false;
+               expect(REMOTEUSER_SET).is.not.null;
+               expect(REMOTEUSER_SET.data.avatarurl).is.null;
+               expect(REMOTEUSER_SET.data.displayname).is.undefined;
+               expect(DISPLAYNAME_SET).is.null;
+           });
+       });
+       it("Will fetch an existing user", () => {
```
Comment (Half-Shot) [inline test/test_usersyncroniser.ts]: This should be titled, "will do nothing"

File: test/test_usersyncroniser.ts [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        },
+    };
+    const config = new DiscordBridgeConfig();
+    config.bridge.domain = "localhost";
+    return new UserSync(bridge as Bridge, config);
+}
+
+describe("UserSyncroniser", () => {
+   describe("GetUserUpdateState", () => {
+       it("Will create new users", () => {
```
Comment (Half-Shot) [inline test/test_usersyncroniser.ts]: Should be titled, will create "a" new user.

File: src/bot.ts [hunk 3]
```diff
@@ -75,13 +82,12 @@ export class DiscordBot {
       if (!this.config.bridge.disablePresence) {
         client.on("presenceUpdate", (_, newMember) => { this.presenceHandler.EnqueueMember(newMember); });
       }
-      client.on("userUpdate", (_, newUser) => { this.UpdateUser(newUser); });
```
Comment (Half-Shot) [inline src/bot.ts]: This shouldn't have been removed.

**Comment by Half-Shot** [comment]
I'd be grateful for a quick pass review from anyone (can I entice you @turt2live or @anoadragon453 ). Don't really feel comfy merging this in as is.

File: src/usersyncroniser.ts [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+            return;
+        }
+        const nickKey = `nick_${guildId}`;
+        const remoteUser = await this.userStore.getRemoteUser(memberState.id);
+        const intent = this.bridge.getIntent(memberState.mxUserId);
+        log.verbose("UserSync", `Sending state event for state ${JSON.stringify(memberState)}.`);
+        await intent.sendStateEvent(roomId, "m.room.member", memberState.mxUserId, {
+            membership: "join",
+            avatar_url: remoteUser.get("avatarurl_mxc"),
+            displayname: memberState.displayName,
```
Comment (turt2live) [inline src/usersyncroniser.ts]: is this PR also meant to handle the guilds stuff? If so, that's missing here.

File: src/matrixroomhandler.ts [hunk 5]
```diff
@@ -89,7 +99,11 @@ export class MatrixRoomHandler {
       return Promise.reject("Event too old");
     }
     if (event.type === "m.room.member" && event.content.membership === "invite") {
-      return this.HandleInvite(event);
+        return this.HandleInvite(event);
+    } else if (event.type === "m.room.member" && event.content.membership === "join") {
+        if (this.bridge.getBot()._isRemoteUser(event.state_key)) {
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/matrixroomhandler.ts]: This one is a sad special case where it should be public, but isn't. I don't really have an alternative function

File: src/usersyncroniser.ts [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+     * @param {module:discord.js.User} New user object
+     * @returns {Promise<void>}
+     * @constructor
+     */
+    public async OnUpdateUser(discordUser: User) {
+        const userState = await this.GetUserUpdateState(discordUser);
+        try {
+            await this.ApplyStateToProfile(userState);
+        } catch (e) {
+            log.error("UserSync", "Failed to update user's profile", e);
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/usersyncroniser.ts]: I chose not to as it's not something the caller usually cares about, if we fail then it just means the user will look a bit ugly for a while. There isn't a recovery step here.

**Pull request merged**